src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,520 @@
+			        HEADER,
+			        text,
+			        this_fn
+			);
+		}
+		break;
+		case UI_SCRIPT_N_OF_K: {
+			{
+				// max possible length 67: "Reuqired signatures: x. Contains x nested scripts"
+				// where x is 2^32-1
+				char text[68];
+				SAFE_SPRINTF(text, "Required signatures: %u. Contains %u nested scripts", ctx->scriptContent.requiredScripts, ctx->complexScripts[ctx->level].remainingScripts);
+
+				ui_displayPaginatedText(
+				        HEADER,
+				        text,
+				        this_fn
+				);
+			}
+			break;
+		}
+		case UI_SCRIPT_INVALID_BEFORE:
+		case UI_SCRIPT_INVALID_HEREAFTER:
+			ui_displayUint64Screen(
+			        HEADER,
+			        ctx->scriptContent.timelock,
+			        this_fn
+			);
+			break;
+		default:
+			THROW(ERR_INVALID_STATE);
+		}
+	}
+
+	UI_STEP(DISPLAY_UI_STEP_RESPOND) {
+		io_send_buf(SUCCESS, NULL, 0);
+		ui_displayBusy();
+	}
+
+	UI_STEP_END(DISPLAY_UI_STEP_INVALID);
+}
+#undef HEADER
+#undef ASSERT_UI_SCRIPT_TYPE_SANITY
+
+#define UI_DISPLAY_SCRIPT(UI_TYPE)\
+	ctx->ui_scriptType = UI_TYPE;\
+	ctx->ui_step = DISPLAY_UI_STEP_POSITION;\
+	deriveScriptHash_display_ui_runStep();
+
+// Start complex native script
+
+static void deriveNativeScriptHash_handleAll(read_view_t* view)
+{
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE_WITH_CTX("");
+
+	nativeScriptHashBuilder_startComplexScript_all(&ctx->hashBuilder, ctx->complexScripts[ctx->level].remainingScripts);
+
+	ctx->complexScripts[ctx->level].ui_type = UI_SCRIPT_ALL;
+	UI_DISPLAY_SCRIPT(UI_SCRIPT_ALL);
+}
+
+static void deriveNativeScriptHash_handleAny(read_view_t* view)
+{
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE_WITH_CTX("");
+
+	nativeScriptHashBuilder_startComplexScript_any(&ctx->hashBuilder, ctx->complexScripts[ctx->level].remainingScripts);
+
+	ctx->complexScripts[ctx->level].ui_type = UI_SCRIPT_ANY;
+	UI_DISPLAY_SCRIPT(UI_SCRIPT_ANY);
+}
+
+static void deriveNativeScriptHash_handleNofK(read_view_t* view)
+{
+	// parse data
+	ctx->scriptContent.requiredScripts = parse_u4be(view);
+
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE_WITH_CTX("required scripts = %u, ", ctx->scriptContent.requiredScripts);
+
+	// validate that the received requiredScripts count makes sense
+	VALIDATE(ctx->complexScripts[ctx->level].remainingScripts >= ctx->scriptContent.requiredScripts, ERR_INVALID_DATA);
+
+	nativeScriptHashBuilder_startComplexScript_n_of_k(&ctx->hashBuilder, ctx->scriptContent.requiredScripts, ctx->complexScripts[ctx->level].remainingScripts);
+
+	ctx->complexScripts[ctx->level].ui_type = UI_SCRIPT_N_OF_K;
+	UI_DISPLAY_SCRIPT(UI_SCRIPT_N_OF_K);
+}
+
+static void deriveNativeScriptHash_handleComplexScriptStart(read_view_t* view)
+{
+	TRACE_BUFFER(view->begin, (size_t) (view->end - view->begin));
+
+	expectsMoreScripts();
+
+	// check if we can increase the level without breaking the MAX_SCRIPT_DEPTH constraint
+	VALIDATE(ctx->level + 1 < MAX_SCRIPT_DEPTH, ERR_DATA_TOO_LARGE);
```

Comment:
The error ERR_DATA_TOO_LARGE is not used consistently throughout the codebase, I'm afraid. The original intention is to report a buffer that is too large; I would suggest using ERR_INVALID_DATA here. (And please create an issue for the code cleanup regarding these two errors: they should get a clearly defined meaning, mentioned in a comment near their definition, and then used throughout the code according to the meaning.)

Reply:
Created the issue https://github.com/vacuumlabs/ledger-app-cardano-shelley/issues/92